### PR TITLE
[DO NOT MERGE] Virtio wip 5 3

### PR DESCRIPTION
**Pull request by lyakh**
This is for an early review only. This version does playback and capture, but it uses ACRN deprecated / proprietary APIs, which in the next step have to be replaced with standard VirtIO API calls. However, most likely that change will only affect virtio-be.c, and virtio-miscdev.c will probably disappear. But the guest / the front end (virtio-fe.c) and any core changes can probably be reviewed already.

### CODE REVIEW

**Comment by plbossart** [comment]
@lyakh this is very difficult to review.
Can I suggest an alternate path, similar to the one we've been using for SoundWire
a) you create a branch that adds all the required ACRN/virtio stuff on top of topic/sof-dev
b) you submit a PR against the branch created in a) so that we only see the commits that are strictly related to SOF.
That way we can provide review feedback on the second part, and when the first is somehow upstreamed or merged we can drop it. And whoever wants to try just needs to merge a) and b).

**Comment by lyakh** [comment]
> @lyakh this is very difficult to review.
> Can I suggest an alternate path, similar to the one we've been using for SoundWire
> a) you create a branch that adds all the required ACRN/virtio stuff on top of topic/sof-dev
> b) you submit a PR against the branch created in a) so that we only see the commits that are strictly related to SOF.
> That way we can provide review feedback on the second part, and when the first is somehow upstreamed or merged we can drop it. And whoever wants to try just needs to merge a) and b).

@plbossart sorry, I forgot to mention, that only the last 32 patches are relevant. Of them 2 aren't mine, I just needed them, many are cosmetic, trivial to review or can be skipped entirely for now. There are really somewhere between 5 and 10 patches that deserve review in this PR.

File: include/uapi/sound/sof/virtio.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+/* first the FE need send empty buffer to BE to get the notification */
+#define SOF_VIRTIO_IPC_NOT_TX_VQ_NAME  "sof-ipc-not-tx"
+
+/* the vq to get the notification */
+#define SOF_VIRTIO_IPC_NOT_RX_VQ_NAME  "sof-ipc-not-rx"
+
+struct sof_vfe_ipc_tplg_req {
+	struct sof_ipc_cmd_hdr hdr;
+	char file_name[64];
```
Comment (lgirdwood) [inline include/uapi/sound/sof/virtio.h]: Can you add comments here, what is this file used for ?

File: sound/soc/intel/boards/bxt_pcm512x.c [hunk 2]
```diff
@@ -166,6 +166,12 @@ static const struct snd_soc_ops aif1_ops = {
 	.shutdown = aif1_shutdown,
 };
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_VIRTIO_BE)
+#define SSP5_NAME "vm_dai_link"
+#else
```
Comment (lgirdwood) [inline sound/soc/intel/boards/bxt_pcm512x.c:171]: We cant do this (conditional compillation), I assume this is for Guests ? GOS machine driver will be very generic and use topology to defne BE and FE DAI links. SOS machine drivers wont change at all.

File: sound/soc/soc-pcm.c [hunk 3]
```diff
@@ -1428,7 +1433,8 @@ static bool dpcm_end_walk_at_be(struct snd_soc_dapm_widget *widget,
 
 	if (dir == SND_SOC_DAPM_DIR_OUT) {
 		for_each_card_rtds(card, rtd) {
-			if (!rtd->dai_link->no_pcm)
+			if (!rtd->dai_link->no_pcm ||
+			    !strcmp(widget->name, SOC_VIRT_DAI_PLAYBACK))
```
Comment (lgirdwood) [inline sound/soc/soc-pcm.c:1437]: Maybe better to use a bool flag here like no_pcm ? This should be a seperate patch too btw.

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -725,6 +725,9 @@ static const struct sof_topology_token comp_tokens[] = {
 	{SOF_TKN_COMP_FORMAT,
 		SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_comp_format,
 		offsetof(struct sof_ipc_comp_config, frame_fmt), 0},
+	{SOF_TKN_COMP_REF,
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c:728]: A lot of this looks like it should be seperate feature/fix patches.

**Comment by plbossart** [comment]
> > @lyakh this is very difficult to review.
> > Can I suggest an alternate path, similar to the one we've been using for SoundWire
> > a) you create a branch that adds all the required ACRN/virtio stuff on top of topic/sof-dev
> > b) you submit a PR against the branch created in a) so that we only see the commits that are strictly related to SOF.
> > That way we can provide review feedback on the second part, and when the first is somehow upstreamed or merged we can drop it. And whoever wants to try just needs to merge a) and b).
> 
> @plbossart sorry, I forgot to mention, that only the last 32 patches are relevant. Of them 2 aren't mine, I just needed them, many are cosmetic, trivial to review or can be skipped entirely for now. There are really somewhere between 5 and 10 patches that deserve review in this PR.

exactly, so please do us a favor so that we can avoid clicking 145 times to see the relevant commits :-)

File: sound/soc/soc-pcm.c [hunk 5]
```diff
@@ -1486,6 +1493,13 @@ static int dpcm_prune_paths(struct snd_soc_pcm_runtime *fe, int stream,
 		/* is there a valid CPU DAI widget for this BE */
 		widget = dai_get_widget(dpcm->be->cpu_dai, stream);
 
+		/*
+		 * The problem with VirtIO is, that as long as .playback_widget
+		 * or .capture_widget respectively is set, the back-end isn't
+		 * pruned. If we force the .{playback,capture}_widget = NULL,
+		 * then later that widget isn't included in the clean up.
+		 */
+
```
Comment (ranj063) [inline sound/soc/soc-pcm.c:1502]: @lyakh there's no code to justify the comment? Is this a TODO?

File: sound/soc/soc-pcm.c [hunk 6]
```diff
@@ -3176,6 +3184,14 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
...
 	}
 
+	if (rtd->dai_link->no_pcm) {
+		if (playback)
+			pcm->streams[SNDRV_PCM_STREAM_PLAYBACK].substream->private_data = rtd;
+		if (capture)
+			pcm->streams[SNDRV_PCM_STREAM_CAPTURE].substream->private_data = rtd;
+		goto out;
+	}
+
```
Comment (ranj063) [inline sound/soc/soc-pcm.c:3194]: @lyakh need a comment here for this change. Why add pcm ops for a dai_link with no_pcm?

File: sound/soc/sof/core.c [hunk 7]
```diff
@@ -319,7 +320,8 @@ static int sof_dsp_start(struct snd_sof_dev *sdev)
 
 fw_run_err:
 #if !IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE)
-	snd_sof_fw_unload(sdev);
+	if (!sdev->is_vfe)
```
Comment (ranj063) [inline sound/soc/sof/core.c:323]: @lyakh where is is_vfe is set?

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -2711,6 +2732,11 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 	if (!link->no_pcm) {
 		link->nonatomic = true;
 
+		if (link->stream_name &&
+		    (!strcmp(link->stream_name, "vm_fe_playback") ||
+		     !strcmp(link->stream_name, "vm_fe_capture")))
+			link->no_pcm = true;
+
```
Comment (ranj063) [inline sound/soc/sof/topology.c:2742]: @lyakh its very hard to guess what is happening here. Can you please add a comment? 
Typically, DAI links for FE's are created by the soc core while parsing pcm elems in the topology during topology parsing . So their names are just dummy names. When do you set the stream_name's get set to vm_fe_playback/capture?

File: sound/soc/sof/topology.c [hunk 9]
```diff
@@ -2894,6 +2920,12 @@ static int sof_route_load(struct snd_soc_component *scomp, int index,
 		goto err;
 	}
 
+	if (source_swidget->id == snd_soc_dapm_buffer) {
+		struct sof_ipc_buffer *buffer = source_swidget->private;
+		if (!buffer->size)
+			goto err;
+	}
+
```
Comment (ranj063) [inline sound/soc/sof/topology.c:2932]: comment here too.

Comment (ranj063) [reply sound/soc/sof/topology.c:2932]: What kind of route does this imply? a virtual VM buffer to?

File: sound/soc/intel/boards/bxt_pcm512x.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/intel/boards/bxt_pcm512x.c:171]: @lgirdwood no, actually that's for the host. "BE" is Back End, i.e. the host / SOS. I have to clean up all the special names that are used throughout the patch set like this one, also "VM FE Playback", "VM FE Capture", "VIO", "vm_fe_playback", "vm_fe_capture" - I have to find a minimum and meaningful set of those

File: sound/soc/soc-pcm.c (hunk 3 above)
Comment (lyakh) [reply sound/soc/soc-pcm.c:1437]: @lgirdwood I'll have to work more on making a meaningful split of the patches, but so far it seems to me, that there will anyway be a relatively big core ALSA / SOF VirtIO patch still. It will contain both the VirtIO (guest) driver itself and supporting code like the above, because the supporting code makes no sense without the driver files, and driver files are non-functional or potentially even crashing without supporting code. Maybe this will become a bool flag in the end, yes.

File: sound/soc/soc-pcm.c (hunk 5 above)
Comment (lyakh) [reply sound/soc/soc-pcm.c:1502]: @ranj063 the comment isn't well placed... No, it isn't a TODO, this has been fixed already, but the comment and the actual fix somehow drifted apart...

File: sound/soc/sof/core.c (hunk 7 above)
Comment (lyakh) [reply sound/soc/sof/core.c:323]: @ranj063 in sof_virtio_vfe_init()

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:2742]: @ranj063 I hard-code those names in the topology for now. We do need to come up with a good consistent scheme for handling of those names / VirtIO detection, etc.

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:2932]: @ranj063 yes, these are virtual buffers of size 0, serving as a handle to the mixer / demux, where the actual pipeline from the guest will be attached. I'll add comments, thanks

**Comment by lyakh** [comment]
The updated version of this PR isn't using any proprietary ACRN calls any more, so, it should now be possible to re-test it with a pure VirtIO / vhost setup like KVM / QEMU and eventually push into the mainline. For this QEMU has to be modified to support the new vhost-dsp device.
Since I haven't been able to get zero-copy to work under ACRN without using ACRN proprietary API, I had to send audio data between the VMs over virtual queues. I would propose to first clean up and merge this solution and then to work on a zero-copy implementation.

**Comment by plbossart** [comment]
@lyakh please submit the improvements separately. I see you have memory leaks that were fixed but not contributed to topic/sof-dev. This is not quite right...

File: sound/soc/sof/core.c [hunk 10]
```diff
@@ -526,4 +526,3 @@ EXPORT_SYMBOL(snd_sof_device_remove);
 MODULE_AUTHOR("Liam Girdwood");
 MODULE_DESCRIPTION("Sound Open Firmware (SOF) Core");
 MODULE_LICENSE("Dual BSD/GPL");
-MODULE_ALIAS("platform:sof-audio");
```
Comment (plbossart) [inline sound/soc/sof/core.c:529]: please submit this directly in another PR

File: sound/soc/sof/intel/hda-codec.c [hunk 11]
```diff
@@ -113,6 +114,9 @@ int hda_codec_probe_bus(struct snd_sof_dev *sdev)
 		}
 	}
 
+	/* set autosuspend delay for hda bus device */
+	snd_hda_set_power_save(hbus, SND_SOF_SUSPEND_DELAY_MS);
+
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-codec.c:119]: we don't want this commit, we agreed not to do this already.

File: sound/soc/soc-topology.c [hunk 12]
```diff
@@ -115,11 +115,7 @@ static int soc_tplg_check_elem_count(struct soc_tplg *tplg, size_t elem_size,
 
 static inline int soc_tplg_is_eof(struct soc_tplg *tplg)
 {
-	const u8 *end = tplg->hdr_pos;
-
-	if (end >= tplg->fw->data + tplg->fw->size)
-		return 1;
-	return 0;
+	return tplg->hdr_pos >= tplg->fw->data + tplg->fw->size;
```
Comment (plbossart) [inline sound/soc/soc-topology.c:118]: please submit this directly in another PR

File: sound/soc/sof/intel/hda-ipc.c [hunk 13]
```diff
@@ -192,33 +220,8 @@ irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
...
-			/* normal message - process normally */
-			snd_sof_ipc_msgs_rx(sdev);
-		}
-
-		hda_dsp_ipc_host_done(sdev);
-
-		ipc_irq = true;
-	}
+	if (host_msg)
+		snd_sof_ipc_msgs_rx(sdev);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c:224]: If this patch is needed, submit it as a separate PR. I don;t see any reason why you keep changes to yourself, this is not related to Virtio,

File: sound/soc/sof/intel/hda-pcm.c [hunk 14]
```diff
@@ -89,13 +89,19 @@ int hda_dsp_pcm_hw_params(struct snd_sof_dev *sdev,
...
 
 	size = params_buffer_bytes(params);
 	rate = get_mult_div(sdev, params_rate(params));
 	bits = get_bits(sdev, params_width(params));
 
+	spcm = snd_sof_find_spcm_dai(sdev, substream->private_data);
+	if (spcm)
+		INIT_WORK(&spcm->stream[substream->stream].period_elapsed_work,
+			  sof_pcm_period_elapsed_work);
+
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c:104]: ok, I stop here.
These changes should either be dropped or submitted to the team.
This development flow is not ok in my book.

**Comment by lyakh** [comment]
> @lyakh this is very difficult to review.
> Can I suggest an alternate path, similar to the one we've been using for SoundWire
> a) you create a branch that adds all the required ACRN/virtio stuff on top of topic/sof-dev
> b) you submit a PR against the branch created in a) so that we only see the commits that are strictly related to SOF.
> That way we can provide review feedback on the second part, and when the first is somehow upstreamed or merged we can drop it. And whoever wants to try just needs to merge a) and b).

@plbossart sorry, I forgot to mention, that only the last 32 patches are relevant. Of them 2 aren't mine, I just needed them, many are cosmetic, trivial to review or can be skipped entirely for now. There are really somewhere between 5 and 10 patches that deserve review in this PR.

> @lyakh I will try to remain polite and respectful but I just completely disagree with this development flow. You have a bunch of patches that are unrelated to virtio, and you don't submit them to the team?
> That's a big no-no.
> 
> Also I did mention multiple times that you can have a virtio branch with all the ACRN/KVM changes on top of topic/sof-dev, and then you submit changes on top. That would allow for a much smaller PR and limit the deltas.

@plbossart This PR was marked as a WiP. My primary goal was to give an early indication of what's cooking under the lid. All the early reviews are appreciated and will be taken into account (at least I'll do my best to do that. If I miss any - that's where you can shoot me). I'm now beginning to unroll all that pile and submit it in multiple PRs - as they obviously should be submitted. I understand, that some of those patches on my stack were indeed improvements and maybe also fixes, but in my understanding none of them were critical. I certainly understand and agree, that ideally I should have submitted them in parallel to my main work, but for that I would have to update my tree to the current sof-dev, verify, that those patches still apply and are relevant, apply / forward-port, test and submit them. Then switch back to my main development branch. Then if those patches get comments, I'd have to repeat the whole process again to update the patches etc. So, unless some of those my patches are indeed critical, I'd rather delay that work until the clean up day, i.e. until now.
Today I've cleaned up my virtualisation work on top of the current sof-dev. I'll test it tomorrow and if it goes well, I'll push it and then also push various improvements.
Sorry for the inconvenience.

File: sound/soc/soc-pcm.c (hunk 6 above)
Comment (lyakh) [reply sound/soc/soc-pcm.c:3194]: I need the .prepare() hook to configure the hardware for the VM operation. I call it when the TRIGGER -> START IPC is executed. I'll add a comment

File: sound/soc/intel/boards/bxt_pcm512x.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/intel/boards/bxt_pcm512x.c:171]: No, this was actually for the host. On one hand, this specific hunk is for the bxt-pcm512x machine driver and as such, this is mostly an example (this will also be a separate patch). On the other hand - yes, this string matching and hard-coded names across the code are ugly. Unfortunately I haven't got the time yet to rework them. Each of them has to be considered and a better solution has to be found. This is one of the present limitations of the patch set.

File: sound/soc/soc-pcm.c (hunk 3 above)
Comment (lyakh) [reply sound/soc/soc-pcm.c:1437]: This is another of those string matching locations.

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:728]: well, we can make it a separate commit, but it's only needed for virtualisation (so far at least), so, as a separate patch motivation would be unclear.

File: sound/soc/sof/core.c (hunk 7 above)
Comment (lyakh) [reply sound/soc/sof/core.c:323]: in the frontend code:
```
	sdev->is_vfe = IS_ENABLED(CONFIG_SND_SOC_SOF_VIRTIO_FE);
```

**Comment by lyakh** [comment]
This will now be split in multiple PRs, of them the main one - the actual virtualisation implementation is #1501. I'm closing this one, but I will use it to track comments.
